Terminal freezes when running my custom action

# What is the peco version, OS, architecture? 

- peco version - 0.4.2
- OS - RHEL 6.6
- Terminal - xterm
- Shell - tcsh
- Using in tmux master branch build

# Abstract

My custom action causes `peco` to freeze and in the process of killing the stuck process, I also lose my terminal cursor. So then I eventually need to kill that terminal too.

# Discussion

I have [this custom action named `my.ToggleSelectionInAboveLine`](https://github.com/kaushalmodi/dotfiles/blob/a41ea9d224850bba3470c4238562bfc7779be8f5/.config/peco/config.json#L121-L124), and I have [bound it to `C-u,C-Space`](https://github.com/kaushalmodi/dotfiles/blob/a41ea9d224850bba3470c4238562bfc7779be8f5/.config/peco/config.json#L139).

When I use the `C-u,S-Space` binding, peco freezes, and I need to use `kill -9` to kill it from another terminal. Then the terminal recovers, but I do not see my regular cursor in the terminal any more!

So there are 2 issues:

1. `peco is freezing due to this action:  

   ```
        "my.ToggleSelectionInAboveLine": [
            "peco.SelectUp",
            "peco.ToggleSelectionAndSelectNext"
        ],
    ```

2. After `peco` is killed, I lose the cursor in my terminal.

  This is what a selected terminal (pane, in tmux) looks like normally:

  - Note that the upper terminal is selected and you can see the cursor in that

  ![image](https://cloud.githubusercontent.com/assets/3578197/18960845/190cfccc-8639-11e6-86d1-f6094e014c02.png)

  The bottom terminal is where I had to kill the peco process. Now even though that is selected, you see no cursor in there.

  ![image](https://cloud.githubusercontent.com/assets/3578197/18960902/46b661f4-8639-11e6-8f2a-4d4d8ded923d.png)


# My config
- [full config code](https://github.com/kaushalmodi/dotfiles/blob/master/.config/peco/config.json)

Bummer, I tried installing `glide` and now I face this issue:

```
km²~/downloads/:git/glide> which go                                                                                 09/30 5:23pm
/home/kmodi/go1.7.1/bin/go
km²~/downloads/:git/glide> make build                                                                               09/30 5:23pm
go build -o glide -ldflags "-X main.version=v0.12.1-16-gbd55e76" glide.go
glide.go:24:2: cannot find package "github.com/Masterminds/glide/action" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/action (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/action (from $GOPATH)
glide.go:25:2: cannot find package "github.com/Masterminds/glide/cache" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/cache (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/cache (from $GOPATH)
glide.go:26:2: cannot find package "github.com/Masterminds/glide/msg" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/msg (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/msg (from $GOPATH)
glide.go:27:2: cannot find package "github.com/Masterminds/glide/path" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/path (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/path (from $GOPATH)
glide.go:28:2: cannot find package "github.com/Masterminds/glide/repo" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/repo (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/repo (from $GOPATH)
glide.go:29:2: cannot find package "github.com/Masterminds/glide/util" in any of:
        /home/kmodi/go1.7.1/src/github.com/Masterminds/glide/util (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/Masterminds/glide/util (from $GOPATH)
glide.go:31:2: cannot find package "github.com/codegangsta/cli" in any of:
        /home/kmodi/go1.7.1/src/github.com/codegangsta/cli (from $GOROOT)
        /home/kmodi/go.apps/src/github.com/codegangsta/cli (from $GOPATH)
make: *** [build] Error 1
```

I also updated to the latest stable release of go: 1.7.1

From the `glide` github page, I followed these steps:

> To build from source you can: 
> 1. Clone this repository into $GOPATH/src/github.com/Masterminds/glide and change directory into it 
> 2. If you are using Go 1.5 ensure the environment variable GO15VENDOREXPERIMENT is set, for example by running export GO15VENDOREXPERIMENT=1. In Go 1.6 it is enabled by default and in Go 1.7 it is always enabled without the ability to turn it off. 
> 3. Run make build

I already got glide installed and it's in my `$PATH`.

Sorry, still no luck with getting the local build working:

Does the `[WARN]` in the below messages mean anything serious?

```
km²~/downloads/:git/peco_orig> make installdeps
Installing glide for linux/amd64...
linux-amd64/
linux-amd64/glide
linux-amd64/LICENSE
linux-amd64/README.md
Installing dependencies...
[WARN] The name listed in the config file (github.com/peco/peco) does not match the current location (.)
[INFO] Downloading dependencies. Please wait...
[INFO] Fetching updates for github.com/google/btree.
[INFO] Fetching updates for github.com/jessevdk/go-flags.
[INFO] Fetching updates for github.com/mattn/go-runewidth.
[INFO] Fetching updates for github.com/nsf/termbox-go.
[INFO] Fetching updates for github.com/pkg/errors.
[INFO] Fetching updates for github.com/davecgh/go-spew.
[INFO] Fetching updates for github.com/lestrrat/go-pdebug.
[INFO] Fetching updates for github.com/stretchr/testify.
[INFO] Fetching updates for golang.org/x/net.
[INFO] Downloading dependencies. Please wait...
[INFO] Setting references.
[INFO] Setting version for github.com/davecgh/go-spew to 5215b55f46b2b919f50a1df0eaa5886afe4e3b3d.
[INFO] Setting version for github.com/google/btree to 00edb8c3163323f673bbe3c04afd9429eb12117d.
[INFO] Setting version for github.com/lestrrat/go-pdebug to 2e6eaaa5717f81bda41d27070d3c966f40a1e75f.
[INFO] Setting version for github.com/nsf/termbox-go to 362329b0aa6447eadd52edd8d660ec1dff470295.
[INFO] Setting version for github.com/stretchr/testify to 6cb3b85ef5a0efef77caef88363ec4d4b5c0976d.
[INFO] Setting version for github.com/pkg/errors to d814416a46cbb066b728cfff58d30a986bc9ddbe.
[INFO] Setting version for github.com/jessevdk/go-flags to 6b9493b3cb60367edd942144879646604089e3f7.
[INFO] Setting version for github.com/mattn/go-runewidth to d6bea18f789704b5f83375793155289da36a3c7f.
[INFO] Setting version for golang.org/x/net to ef00b378c73f107bf44d5c9b69875255ce89b79a.
km²~/downloads/:git/peco_orig> make test                                                                            09/30 5:57pm
Installing dependencies...
[WARN] The name listed in the config file (github.com/peco/peco) does not match the current location (.)
[INFO] Downloading dependencies. Please wait...
[INFO] Fetching updates for github.com/jessevdk/go-flags.
[INFO] Fetching updates for github.com/lestrrat/go-pdebug.
[INFO] Fetching updates for github.com/nsf/termbox-go.
[INFO] Fetching updates for github.com/pkg/errors.
[INFO] Fetching updates for github.com/stretchr/testify.
[INFO] Fetching updates for github.com/davecgh/go-spew.
[INFO] Fetching updates for github.com/mattn/go-runewidth.
[INFO] Fetching updates for golang.org/x/net.
[INFO] Fetching updates for github.com/google/btree.
[INFO] Downloading dependencies. Please wait...
[INFO] Setting references.
[INFO] Setting version for github.com/mattn/go-runewidth to d6bea18f789704b5f83375793155289da36a3c7f.
[INFO] Setting version for github.com/lestrrat/go-pdebug to 2e6eaaa5717f81bda41d27070d3c966f40a1e75f.
[INFO] Setting version for github.com/nsf/termbox-go to 362329b0aa6447eadd52edd8d660ec1dff470295.
[INFO] Setting version for github.com/pkg/errors to d814416a46cbb066b728cfff58d30a986bc9ddbe.
[INFO] Setting version for github.com/google/btree to 00edb8c3163323f673bbe3c04afd9429eb12117d.
[INFO] Setting version for github.com/stretchr/testify to 6cb3b85ef5a0efef77caef88363ec4d4b5c0976d.
[INFO] Setting version for github.com/jessevdk/go-flags to 6b9493b3cb60367edd942144879646604089e3f7.
[INFO] Setting version for github.com/davecgh/go-spew to 5215b55f46b2b919f50a1df0eaa5886afe4e3b3d.
[INFO] Setting version for golang.org/x/net to ef00b378c73f107bf44d5c9b69875255ce89b79a.
Running tests...
package ./cmd/peco
        imports github.com/peco/peco/internal/util: use of internal package not allowed
package .
        imports github.com/peco/peco/internal/keyseq: use of internal package not allowed
package .
        imports github.com/peco/peco/internal/util: use of internal package not allowed
make: *** [test] Error 1
```

--- a/action.go
+++ b/action.go
@@ -324,6 +324,10 @@ func doSelectDown(ctx context.Context, state *Peco, e termbox.Event) {
 }
 
 func doSelectUp(ctx context.Context, state *Peco, e termbox.Event) {
+	if pdebug.Enabled {
+		g := pdebug.Marker("doSelectUp")
+		defer g.End()
+	}
 	state.Hub().SendPaging(ToLineAbove)
 }
 
@@ -344,7 +348,9 @@ func doScrollRight(ctx context.Context, state *Peco, e termbox.Event) {
 }
 
 func doToggleSelectionAndSelectNext(ctx context.Context, state *Peco, e termbox.Event) {
+	toplevel, _ := ctx.Value(isTopLevelActionCall).(bool)
 	state.Hub().Batch(func() {
+		ctx = context.WithValue(ctx, isTopLevelActionCall, false)
 		doToggleSelection(ctx, state, e)
 		// XXX This is sucky. Fix later
 		if state.LayoutType() == "top-down" {
@@ -352,7 +358,7 @@ func doToggleSelectionAndSelectNext(ctx context.Context, state *Peco, e termbox.
 		} else {
 			doSelectUp(ctx, state, e)
 		}
-	})
+	}, toplevel)
 }
 
 func doInvertSelection(ctx context.Context, state *Peco, _ termbox.Event) {
@@ -652,6 +658,11 @@ func doRefreshScreen(ctx context.Context, state *Peco, _ termbox.Event) {
 }
 
 func doToggleQuery(ctx context.Context, state *Peco, _ termbox.Event) {
+	if pdebug.Enabled {
+		g := pdebug.Marker("doToggleQuery")
+		defer g.End()
+	}
+
 	q := state.Query()
 	if q.Len() == 0 {
 		q.RestoreSavedQuery()
@@ -688,19 +699,23 @@ func doSingleKeyJump(ctx context.Context, state *Peco, e termbox.Event) {
 		return
 	}
 
+	toplevel, _ := ctx.Value(isTopLevelActionCall).(bool)
 	state.Hub().Batch(func() {
+		ctx = context.WithValue(ctx, isTopLevelActionCall, false)
 		state.Hub().SendPaging(JumpToLineRequest(index))
 		doFinish(ctx, state, e)
-	})
+	}, toplevel)
 }
 
 func makeCombinedAction(actions ...Action) ActionFunc {
 	return ActionFunc(func(ctx context.Context, state *Peco, e termbox.Event) {
+		toplevel, _ := ctx.Value(isTopLevelActionCall).(bool)
 		state.Hub().Batch(func() {
+			ctx = context.WithValue(ctx, isTopLevelActionCall, false)
 			for _, a := range actions {
 				a.Execute(ctx, state, e)
 			}
-		})
+		}, toplevel)
 	})
 }
 

--- a/hub/hub.go
+++ b/hub/hub.go
@@ -39,10 +39,16 @@ func New(bufsiz int) *Hub {
 
 // Batch allows you to synchronously send messages during the
 // scope of f() being executed.
-func (h *Hub) Batch(f func()) {
-	// lock during this operation
-	h.mutex.Lock()
-	defer h.mutex.Unlock()
+func (h *Hub) Batch(f func(), shouldLock bool) {
+	if pdebug.Enabled {
+		g := pdebug.Marker("Batch %t", shouldLock)
+		defer g.End()
+	}
+	if shouldLock {
+		// lock during this operation
+		h.mutex.Lock()
+		defer h.mutex.Unlock()
+	}
 
 	// temporarily set isSync = true
 	o := h.isSync

--- a/hub/hub_test.go
+++ b/hub/hub_test.go
@@ -50,7 +50,7 @@ func TestHub(t *testing.T) {
 		h.SendDraw(true)
 		h.SendStatusMsg("Hello, World!")
 		h.SendPaging(1)
-	})
+	}, true)
 
 	phases := []string{
 		"query",

--- a/issues_test.go
+++ b/issues_test.go
@@ -5,7 +5,9 @@ import (
 	"io/ioutil"
 	"os"
 	"testing"
+	"time"
 
+	termbox "github.com/nsf/termbox-go"
 	"github.com/stretchr/testify/assert"
 	"golang.org/x/net/context"
 )
@@ -67,4 +69,46 @@ func TestIssue212_SanityCheck(t *testing.T) {
 	if !assert.Equal(t, state.config.Layout, "bottom-up", "Default layout type should be 'bottom-up', got '%s'", state.config.Layout) {
 		return
 	}
+}
+
+func TestIssue345(t *testing.T) {
+	cfg, err := newConfig(`{
+	"Keymap": {
+    "C-t": "my.ToggleSelectionInAboveLine"
+	},
+	"Action": {
+		"my.ToggleSelectionInAboveLine": [
+			"peco.SelectUp",
+			"peco.ToggleSelectionAndSelectNext"
+		]
+	}
+}`)
+	if !assert.NoError(t, err, "newConfig should succeed") {
+		return
+	}
+	defer os.Remove(cfg)
+
+	state := newPeco()
+	if !assert.NoError(t, state.config.Init(), "Config.Init should succeed") {
+		return
+	}
+
+	state.Argv = append(state.Argv, []string{"--rcfile", cfg}...)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go state.Run(ctx)
+	defer cancel()
+
+	<-state.Ready()
+
+	ev := termbox.Event{
+		Type: termbox.EventKey,
+		Key:  termbox.KeyCtrlT,
+	}
+	if !assert.NoError(t, state.Keymap().ExecuteAction(ctx, state, ev), "ExecuteAction should succeed") {
+		return
+	}
+
+	time.Sleep(time.Second)
+
 }

--- a/keymap.go
+++ b/keymap.go
@@ -25,13 +25,21 @@ func (km Keymap) Sequence() Keyseq {
 	return km.seq
 }
 
-func (km Keymap) ExecuteAction(ctx context.Context, state *Peco, ev termbox.Event) error {
+const isTopLevelActionCall = "peco.isTopLevelActionCall"
+
+func (km Keymap) ExecuteAction(ctx context.Context, state *Peco, ev termbox.Event) (err error) {
+	if pdebug.Enabled {
+		g := pdebug.Marker("Keymap.ExecuteAction %v", ev).BindError(&err)
+		defer g.End()
+	}
+
 	a := km.LookupAction(ev)
 	if a == nil {
 		return errors.New("action not found")
 	}
 
-	a.(Action).Execute(ctx, state, ev)
+	ctx = context.WithValue(ctx, isTopLevelActionCall, true)
+	a.Execute(ctx, state, ev)
 	return nil
 }
 

--- a/peco.go
+++ b/peco.go
@@ -266,6 +266,10 @@ func (p *Peco) Run(ctx context.Context) (err error) {
 		defer g.End()
 	}
 
+	// do this only once
+	var readyOnce sync.Once
+	defer readyOnce.Do(func() { close(p.readyCh) })
+
 	if err := p.Setup(); err != nil {
 		return errors.Wrap(err, "failed to setup peco")
 	}
@@ -342,7 +346,7 @@ func (p *Peco) Run(ctx context.Context) (err error) {
 		}()
 	}
 
-	close(p.readyCh)
+	readyOnce.Do(func() { close(p.readyCh) })
 
 	// This has tobe AFTER close(p.readyCh), otherwise the query is
 	// ignored by us (queries are not run until peco thinks it's ready)

--- a/peco_test.go
+++ b/peco_test.go
@@ -3,6 +3,8 @@ package peco
 import (
 	"bytes"
 	"fmt"
+	"io"
+	"io/ioutil"
 	"runtime"
 	"sync"
 	"testing"
@@ -44,6 +46,17 @@ func (i *interceptor) record(name string, args []interface{}) {
 	}
 
 	events[name] = append(v, interceptorArgs(args))
+}
+
+func newConfig(s string) (string, error) {
+	f, err := ioutil.TempFile("", "peco-test-config-")
+	if err != nil {
+		return "", err
+	}
+
+	io.WriteString(f, s)
+	f.Close()
+	return f.Name(), nil
 }
 
 func newPeco() *Peco {

